File: sdk/java/com/netease/open/libpoco/sdk/DefaultMatcher.java
package com.netease.open.libpoco.sdk;

import org.json.JSONArray;
import org.json.JSONException;

import java.util.HashMap;
import java.util.Map;

/**
 * Created by adolli on 2017/7/13.
 */

public class DefaultMatcher implements IMatcher {
    private Map<String, Comparator> predicates = new HashMap<>();

    public DefaultMatcher() {
        predicates.put("attr=", new EqualizationComparator());
        predicates.put("attr.*=", new RegexpComparator());
    }

    public boolean match(JSONArray cond, INode node) throws JSONException {
        String op = cond.getString(0);
        JSONArray args = cond.getJSONArray(1);

        if (op.equals("or")) {
            for (int i = 0; i < args.length(); i++) {
                JSONArray arg = args.getJSONArray(i);
                if (this.match(arg, node)) {
                    return true;
                }
            }
            return false;
        }

        if (op.equals("and")) {
            for (int i = 0; i < args.length(); i++) {
                JSONArray arg = args.getJSONArray(i);
                if (!this.match(arg, node)) {
                    return false;
                }
            }
            return true;
        }

        Comparator comparator = this.predicates.get(op);
        if (comparator != null) {
            String attribute = args.getString(0);
            Object value = args.get(1);
            Object targetValue = node.getAttr(attribute);
            return comparator.compare(targetValue, value);
        }

        return false;
    }

    public interface Comparator {
        boolean compare(Object l, Object r);
    }

    public class EqualizationComparator implements Comparator {

        @Override
        public boolean compare(Object l, Object r) {
            if (l == null) {
                return false;
            }
            return l.equals(r);
        }
    }

    public class RegexpComparator implements Comparator {

        @Override
        public boolean compare(Object origin, Object pattern) {
            if (origin == null) {
                return false;
            }
            return ((String) origin).matches((String) pattern);
        }
    }
}

File: cocos-creator/Poco/sdk/DefaultMatcher.js
var IMatcher = function () {
}

IMatcher.prototype.match = function (cond, node) {
    // :rettype: bool
}


var EqualizationComparator = function () {}
EqualizationComparator.prototype.compare = function (l, r) {
    // 现在只判断字符串、数字、bool值的这种简单属性
    // 如果要判断数组或者dict等复杂结构再加
    return l === r
}


var RegexpComparator = function () {}
RegexpComparator.prototype.compare = function (origin, pattern) {
    if (!origin || !pattern) {
        return false
    }
    return origin.toString().match(pattern) !== null
}


var DefaultMatcher = function () {
    IMatcher.call(this)
    this.comparators = {
        'attr=': new EqualizationComparator(),
        'attr.*=': new RegexpComparator(),
    }
}
DefaultMatcher.prototype = Object.create(IMatcher.prototype)


DefaultMatcher.prototype.match = function (cond, node) {
    var op = cond[0]
    var args = cond[1]

    // 条件匹配
    if (op === 'and') {
        for (var i in args) {
            var arg = args[i]
            if (!this.match(arg, node)) {
                return false
            }
        }
        return true
    }

    if (op === 'or') {
        for (var i in args) {
            var arg = args[i]
            if (this.match(arg, node)) {
                return true
            }
        }
        return false
    }

    // 属性匹配
    var comparator = this.comparators[op]
    if (comparator) {
        var attribute = args[0]
        var value = args[1]
        var targetValue = node.getAttr(attribute)
        return comparator.compare(targetValue, value) 
    }

    return false
}

try {
    module.exports = DefaultMatcher;
} catch (e) {
    if (window.module && window.module.exports) {
        window.module.exports = DefaultMatcher;
    }
}
